packages/kbn-custom-events/src/custom_events.ts

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { PackageInfo } from '@kbn/config';
+import { isArray, isInteger, mapKeys } from 'lodash';
+import { FullStoryApi } from './fullstory';
+
+export interface CustomEventConfig {
+  enabled: boolean;
+  fullstoryOrgId?: string;
+  esOrgId?: string;
+  basePath: any;
+  userIdPromise: Promise<string | undefined>;
+  packageInfo: PackageInfo;
+}
+
+// based on https://help.fullstory.com/hc/en-us/articles/360020623234#Custom%20Event%20Name%20Requirements
+type FULLSTORY_EVENT_TYPES = string | number | boolean | Date;
+type FullstoryEventValue = FULLSTORY_EVENT_TYPES | FULLSTORY_EVENT_TYPES[] | undefined;
+
+export class CustomEvents {
+  private fullStory?: FullStoryApi;
```

Comment:
nit: I would find a way to make FullStory an implementation detail, abstracted away as much as possible from the code.  It still feels really baked in, which may or may not be a problem.